src/v/cluster/node_status_provider.h

```diff
@@ -0,0 +1,101 @@
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "cluster/members_table.h"
+#include "cluster/node_status_provider_rpc_service.h"
+#include "config/property.h"
+#include "rpc/connection_cache.h"
+#include "seastarx.h"
+#include "ssx/metrics.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/timer.hh>
+
+#include <absl/container/flat_hash_set.h>
+
+namespace cluster {
+
+struct node_status {
+    model::node_id node_id;
+    rpc::clock_type::time_point last_seen;
+};
+
+/*
+ * node_status_provider is a sub-system that operates outside of Raft
+ * and has the purpose of tracking the status of all peers. Its operation is as
+ * follows:
+ * 1. Maintain a list of peers for this node. This is currently done via a
+ * callback from the members_table.
+ * 2. Send a periodic node_status RPC to all known peers and record the results.
+ *
+ * It runs only on shard 0.
+ */
+class node_status_provider {
+public:
+    static constexpr ss::shard_id shard = 0;
+
+    node_status_provider(
+      model::node_id,
+      ss::sharded<members_table>&,
+      config::binding<std::chrono::milliseconds>);
+
+    void start();
+    ss::future<> stop();
+
+    std::optional<node_status> get_node_status(model::node_id) const;
+
+private:
+    void handle_members_updated_notification(std::vector<model::node_id>);
+
+    void update_node_status(node_status_metadata&&);
+
+    void tick();
+    ss::future<> dispatch_node_status_requests();
+
+    void process_reply(result<node_status_reply>);
+    ss::future<node_status_reply> process_request(node_status_request&&);
+
+    ss::future<> do_node_status(model::node_id, node_status_request);
+    ss::future<result<node_status_reply>>
+      send_node_status_request(model::node_id, node_status_request);
+
+    void setup_metrics(ss::metrics::metric_groups&);
+
+    struct statistics {
+        int64_t rpcs_sent;
+        int64_t rpcs_timed_out;
+        int64_t rpcs_received;
+    };
+
+private:
+    model::node_id _self;
+    ss::sharded<members_table>& _members_table;
+    config::binding<std::chrono::milliseconds> _period;
+    config::tls_config _rpc_tls_config;
+
+    rpc::connection_cache _node_connection_cache;
+    absl::flat_hash_set<model::node_id> _discovered_peers;
+    absl::flat_hash_map<model::node_id, node_status> _peers_state;
+    ss::gate _gate;
+    ss::timer<ss::lowres_clock> _timer;
+    notification_id_type _members_table_notification_handle;
+
+    statistics _stats{};
+    ss::metrics::metric_groups _metrics;
+    ss::metrics::metric_groups _public_metrics{
+      ssx::metrics::public_metrics_handle};
+
+    friend class node_status_provider_rpc_handler;
```

Comment:
this isn't defined in this commit

Reply:
The commit order didn't make any sense. Reordered them and fixed this too.